### ISSUE TITLE
OpenNI callback gets deregistered when Qt GUI is initiated

### ISSUE THREAD

**Issue by indianajohn** [PointCloudLibrary/pcl#1028]
Hi,

I have tested this issue on PCL 1.7.2 and also the most recent Github release on Ubuntu 14.04 (PCL 1.7.2 from the package system and also custom compiled).  I tried linking against both VTK 5.x, 6.x, and the latest Gthub version, as well as Qt 4.x and 5.x. I have also tested this on PCL 1.7.2 on Windows.

Whenever A Qt GUI is started, all cloud callbacks seem to become deregistered.  This problem can be reproduced by starting from visualizer:

https://github.com/PointCloudLibrary/pcl/tree/master/doc/tutorials/content/sources/qt_visualizer

adding a callback function to main.h, starting it, sleeping for a bit, and starting the GUI:

``` C++
#include "pclviewer.h"
#include <QApplication>
#include <QMainWindow>
#include <pcl/io/openni_grabber.h>

int g_nCallbackCount;

void cloud_cb_ (const pcl::PointCloud<pcl::PointXYZA>::ConstPtr &cloud)
{
std::cout << "Hello from the callback" << g_nCallbackCount << std::endl;
g_nCallbackCount++;
}

int main (int argc, char *argv[])
{
// Register callback
pcl::Grabber* interface = new pcl::OpenNIGrabber();
g_nCallbackCount = 0;
boost::function<void (const pcl::PointCloud<pcl::PointXYZ>::ConstPtr&)> f =
boost::bind (&cloud_cb_);
interface->registerCallback (f);
interface->start ();

// Wait a bit
sleep(5);

// Start the GUI
QApplication a (argc, argv);
PCLViewer w;
w.show ();
interface->stop ();
delete interface;
return a.exec ();
}
```

The program starts, calls the callback for 5 seconds, and then the callback dies as the GUI comes up.

I suspect this has to do with namespace collision between boost::signals2 and Qt signals.
http://www.boost.org/doc/libs/1_55_0/doc/html/signals/s04.html

I would appreciate any help with this as I am trying to make a Qt GUI for a program that accesses a Xtion sensor and it's impossible with this bug present.

Thank you.

Best Regards,
John

**Comment by 67shafar** [comment]
Hello,

I know this is late but I had the exact same problem just now. Your intuition was correct. 
You can fix this by adding:
 ADD_DEFINITIONS(-DQT_NO_KEYWORDS) 

to your cmakelists.txt file. You will also want to make sure it comes before ${QT_DEFINITIONS} if that exists in your cmakelists.txt file.

In my case I have this:
ADD_DEFINITIONS (-DQT_NO_KEYWORDS ${QT_DEFINITIONS})

**Comment by indianajohn** [comment]
Thanks for the solution.. it actually helps my current work. I'll leave this open for the time being because this is a pretty baffling pitfall and maybe this flag should be added such that it automatically makes its way into programs linked against PCL.

**Comment by taketwo** [comment]
I think we should convert to use `Q_SLOTS` and `Q_SIGNALS` instead of keywords everywhere in PCL to avoid possible conflicts. For example, [this has been done](https://bugs.webkit.org/show_bug.cgi?id=93996) in WebKit.

**Comment by taketwo** [comment]
I could not reproduce this issue with latest PCL, Qt5, and OpenNI2 grabber. Callback continues to fire after the GUI shows up. Nevertheless, I've submitted a patch converting to Q_* since it is generally a good idea not to use these keywords.

**Comment by SergioRAgostinho** [comment]
Should be closed with #1898. Please reopen if necessary.
